File: sdk/ruby/lib/anyvali/schemas/number_schema.rb
# frozen_string_literal: true

module AnyVali
  class NumberSchema < Schema
    FLOAT_KINDS = %w[number float32 float64].freeze
    INT_KINDS = %w[int int8 int16 int32 int64 uint8 uint16 uint32 uint64].freeze

    INT_RANGES = {
      "int8"   => [-128, 127],
      "int16"  => [-32_768, 32_767],
      "int32"  => [-2_147_483_648, 2_147_483_647],
      "int64"  => [-9_223_372_036_854_775_808, 9_223_372_036_854_775_807],
      "uint8"  => [0, 255],
      "uint16" => [0, 65_535],
      "uint32" => [0, 4_294_967_295],
      "uint64" => [0, 18_446_744_073_709_551_615],
      "int"    => [-9_223_372_036_854_775_808, 9_223_372_036_854_775_807]
    }.freeze

    FLOAT32_MAX = 3.4028235e+38

    def initialize(kind: "number", constraints: {}, **kwargs)
      super(kind: kind, constraints: constraints, **kwargs)
    end

    def min(n)
      dup_with(constraints: @constraints.merge("min" => n))
    end

    def max(n)
      dup_with(constraints: @constraints.merge("max" => n))
    end

    def exclusive_min(n)
      dup_with(constraints: @constraints.merge("exclusiveMin" => n))
    end

    def exclusive_max(n)
      dup_with(constraints: @constraints.merge("exclusiveMax" => n))
    end

    def multiple_of(n)
      dup_with(constraints: @constraints.merge("multipleOf" => n))
    end

    protected

    def validate(value, path, issues, context)
      int_kind = INT_KINDS.include?(@kind)
      float_kind = FLOAT_KINDS.include?(@kind)

      if int_kind
        validate_int(value, path, issues)
      elsif float_kind
        validate_float(value, path, issues)
      end
    end

    private

    def validate_int(value, path, issues)
      # Must be numeric
      unless value.is_a?(Integer) || value.is_a?(Float)
        issues << ValidationIssue.new(
          code: IssueCodes::INVALID_TYPE,
          path: path,
          expected: @kind,
          received: Schema.type_name(value)
        )
        return
      end

      # Must be a whole number
      if value.is_a?(Float) && value != value.floor
        issues << ValidationIssue.new(
          code: IssueCodes::INVALID_TYPE,
          path: path,
          expected: @kind,
          received: "number"
        )
        return
      end

      int_val = value.is_a?(Float) ? value.to_i : value

      # Width range check
      if INT_RANGES.key?(@kind)
        min_val, max_val = INT_RANGES[@kind]
        if int_val < min_val
          issues << ValidationIssue.new(
            code: IssueCodes::TOO_SMALL,
            path: path,
            expected: @kind,
            received: int_val.to_s
          )
          return
        end
        if int_val > max_val
          issues << ValidationIssue.new(
            code: IssueCodes::TOO_LARGE,
            path: path,
            expected: @kind,
            received: int_val.to_s
          )
          return
        end
      end

      validate_numeric_constraints(int_val, path, issues)
    end

    def validate_float(value, path, issues)
      unless value.is_a?(Numeric) && !value.is_a?(Complex)
        issues << ValidationIssue.new(
          code: IssueCodes::INVALID_TYPE,
          path: path,
          expected: @kind,
          received: Schema.type_name(value)
        )
        return
      end

      # For booleans (true/false are not Numeric in Ruby, so this is fine)
      if value == true || value == false
        issues << ValidationIssue.new(
          code: IssueCodes::INVALID_TYPE,
          path: path,
          expected: @kind,
          received: "boolean"
        )
        return
      end

      # float32 range check
      if @kind == "float32"
        fval = value.to_f
        if fval.abs > FLOAT32_MAX && !fval.infinite?
          issues << ValidationIssue.new(
            code: IssueCodes::TOO_LARGE,
            path: path,
            expected: "float32",
            received: value.to_s
          )
          return
        end
      end

      validate_numeric_constraints(value, path, issues)
    end

    def validate_numeric_constraints(value, path, issues)
      if @constraints["min"] && value < @constraints["min"]
        issues << ValidationIssue.new(
          code: IssueCodes::TOO_SMALL,
          path: path,
          expected: @constraints["min"].to_s,
          received: value.to_s
        )
      end

      if @constraints["max"] && value > @constraints["max"]
        issues << ValidationIssue.new(
          code: IssueCodes::TOO_LARGE,
          path: path,
          expected: @constraints["max"].to_s,
          received: value.to_s
        )
      end

      if @constraints["exclusiveMin"] && value <= @constraints["exclusiveMin"]
        issues << ValidationIssue.new(
          code: IssueCodes::TOO_SMALL,
          path: path,
          expected: @constraints["exclusiveMin"].to_s,
          received: value.to_s
        )
      end

      if @constraints["exclusiveMax"] && value >= @constraints["exclusiveMax"]
        issues << ValidationIssue.new(
          code: IssueCodes::TOO_LARGE,
          path: path,
          expected: @constraints["exclusiveMax"].to_s,
          received: value.to_s
        )
      end

      if @constraints["multipleOf"]
        divisor = @constraints["multipleOf"]
        remainder = value.to_f % divisor.to_f
        unless remainder.abs < 1e-10 || (divisor.to_f - remainder.abs).abs < 1e-10
          issues << ValidationIssue.new(
            code: IssueCodes::INVALID_NUMBER,
            path: path,
            expected: divisor.to_s,
            received: value.to_s
          )
        end
      end
    end

    def dup_with(**overrides)
      attrs = {
        kind: @kind,
        constraints: @constraints,
        coerce_config: @coerce_config,
        default_value: @default_value,
        has_default: @has_default,
        custom_validators: @custom_validators
      }.merge(overrides)
      self.class.new(**attrs)
    end
  end
end

File: sdk/csharp/src/AnyVali/Schemas/NumberSchema.cs
namespace AnyVali.Schemas;

public class NumberSchema : Schema<double>
{
    protected string Kind;
    protected object? _min;
    protected object? _max;
    protected object? _exclusiveMin;
    protected object? _exclusiveMax;
    protected object? _multipleOf;

    public NumberSchema(string kind = "number")
    {
        Kind = kind;
    }

    internal override string GetCoercionTarget() => Kind;

    public NumberSchema Min(double n) => WithConstraint("min", n);
    public NumberSchema Max(double n) => WithConstraint("max", n);
    public NumberSchema ExclusiveMin(double n) => WithConstraint("exclusiveMin", n);
    public NumberSchema ExclusiveMax(double n) => WithConstraint("exclusiveMax", n);
    public NumberSchema MultipleOf(double n) => WithConstraint("multipleOf", n);

    internal NumberSchema WithConstraint(string name, object value)
    {
        if (!IsFiniteNumber(value) || name == "multipleOf" && ToDouble(value) <= 0)
            throw new InvalidOperationException($"Invalid numeric constraint: {name}");
        var clone = (NumberSchema)Clone();
        switch (name)
        {
            case "min": clone._min = value; break;
            case "max": clone._max = value; break;
            case "exclusiveMin": clone._exclusiveMin = value; break;
            case "exclusiveMax": clone._exclusiveMax = value; break;
            case "multipleOf": clone._multipleOf = value; break;
            default: throw new InvalidOperationException($"Unknown numeric constraint: {name}");
        }
        return clone;
    }

    public new NumberSchema Default(object? value) => (NumberSchema)base.Default(value);
    public new NumberSchema Coerce(Parse.CoercionConfig? config = null) => (NumberSchema)base.Coerce(config);

    internal override object? Validate(object? input, ValidationContext ctx)
    {
        if (!IsFiniteNumber(input))
        {
            ctx.Issues.Add(new ValidationIssue
            {
                Code = IssueCodes.InvalidType,
                Message = $"Expected {Kind}, received {DescribeType(input)}",
                Path = ctx.ClonePath(),
                Expected = Kind,
                Received = DescribeType(input),
            });
            return null;
        }

        var val = ToDouble(input);

        // float32 MUST reject values outside the binary32 representable range
        // (spec 1.4). Without this, float32 silently accepts any double value,
        // defeating the narrowing guarantee.
        if (Kind == "float32" && val != 0.0 && Math.Abs(val) > Float32Max)
        {
            ctx.Issues.Add(new ValidationIssue
            {
                Code = IssueCodes.TooLarge,
                Message = $"Value {FormatNum(val)} is outside the float32 range",
                Path = ctx.ClonePath(),
                Expected = "float32",
                Received = FormatNum(val),
            });
            return null;
        }

        ValidateConstraints(val, ctx);
        return val;
    }

    private const double Float32Max = 3.4028234663852886e38;

    protected void ValidateConstraints(double val, ValidationContext ctx)
    {
        if (_min is not null && val < ToDouble(_min))
        {
            ctx.Issues.Add(new ValidationIssue
            {
                Code = IssueCodes.TooSmall,
                Message = $"Number must be >= {FormatNum(ToDouble(_min))}",
                Path = ctx.ClonePath(),
                Expected = FormatNum(ToDouble(_min)),
                Received = FormatNum(val),
            });
        }

        if (_max is not null && val > ToDouble(_max))
        {
            ctx.Issues.Add(new ValidationIssue
            {
                Code = IssueCodes.TooLarge,
                Message = $"Number must be <= {FormatNum(ToDouble(_max))}",
                Path = ctx.ClonePath(),
                Expected = FormatNum(ToDouble(_max)),
                Received = FormatNum(val),
            });
        }

        if (_exclusiveMin is not null && val <= ToDouble(_exclusiveMin))
        {
            ctx.Issues.Add(new ValidationIssue
            {
                Code = IssueCodes.TooSmall,
                Message = $"Number must be > {FormatNum(ToDouble(_exclusiveMin))}",
                Path = ctx.ClonePath(),
                Expected = FormatNum(ToDouble(_exclusiveMin)),
                Received = FormatNum(val),
            });
        }

        if (_exclusiveMax is not null && val >= ToDouble(_exclusiveMax))
        {
            ctx.Issues.Add(new ValidationIssue
            {
                Code = IssueCodes.TooLarge,
                Message = $"Number must be < {FormatNum(ToDouble(_exclusiveMax))}",
                Path = ctx.ClonePath(),
                Expected = FormatNum(ToDouble(_exclusiveMax)),
                Received = FormatNum(val),
            });
        }

        if (_multipleOf is not null)
        {
            var remainder = val % ToDouble(_multipleOf);
            if (Math.Abs(remainder) > 1e-10 && Math.Abs(remainder - ToDouble(_multipleOf)) > 1e-10)
            {
                ctx.Issues.Add(new ValidationIssue
                {
                    Code = IssueCodes.InvalidNumber,
                    Message = $"Number must be a multiple of {FormatNum(ToDouble(_multipleOf))}",
                    Path = ctx.ClonePath(),
                    Expected = FormatNum(ToDouble(_multipleOf)),
                    Received = FormatNum(val),
                });
            }
        }
    }

    internal override Dictionary<string, object?> ToNode()
    {
        var node = new Dictionary<string, object?> { ["kind"] = Kind };
        if (_min is not null) node["min"] = _min;
        if (_max is not null) node["max"] = _max;
        if (_exclusiveMin is not null) node["exclusiveMin"] = _exclusiveMin;
        if (_exclusiveMax is not null) node["exclusiveMax"] = _exclusiveMax;
        if (_multipleOf is not null) node["multipleOf"] = _multipleOf;
        AddDefaultAndCoercion(node);
        return node;
    }

    internal override Schema Clone()
    {
        return new NumberSchema(Kind)
        {
            _min = _min, _max = _max,
            _exclusiveMin = _exclusiveMin, _exclusiveMax = _exclusiveMax,
            _multipleOf = _multipleOf,
            DefaultValue = DefaultValue, CoercionCfg = CoercionCfg,
            IsPortable = IsPortable, MetadataMap = MetadataMap, ImportedDefinitions = ImportedDefinitions, ImportedExtensions = ImportedExtensions,
        };
    }

    internal static string FormatNum(double d)
    {
        if (d == Math.Floor(d) && !double.IsInfinity(d))
            return ((long)d).ToString();
        return d.ToString("G");
    }
}

public sealed class Float32Schema : NumberSchema
{
    public Float32Schema() : base("float32") { }
    internal override Schema Clone() => new Float32Schema
    {
        _min = _min, _max = _max, _exclusiveMin = _exclusiveMin,
        _exclusiveMax = _exclusiveMax, _multipleOf = _multipleOf,
        DefaultValue = DefaultValue, CoercionCfg = CoercionCfg, IsPortable = IsPortable, MetadataMap = MetadataMap, ImportedDefinitions = ImportedDefinitions, ImportedExtensions = ImportedExtensions,
    };
}

public sealed class Float64Schema : NumberSchema
{
    public Float64Schema() : base("float64") { }
    internal override Schema Clone() => new Float64Schema
    {
        _min = _min, _max = _max, _exclusiveMin = _exclusiveMin,
        _exclusiveMax = _exclusiveMax, _multipleOf = _multipleOf,
        DefaultValue = DefaultValue, CoercionCfg = CoercionCfg, IsPortable = IsPortable, MetadataMap = MetadataMap, ImportedDefinitions = ImportedDefinitions, ImportedExtensions = ImportedExtensions,
    };
}
